test/tls-provider.c

```diff
@@ -451,14 +872,51 @@ static const OSSL_ALGORITHM tls_prov_kem[] = {
 
 /* Key Management for the dummy XOR key exchange algorithm */
 
-static void *xor_newdata(void *provctx)
+static void *xor_newkey(void *provctx)
+{
+    XORKEY* ret = OPENSSL_zalloc(sizeof(XORKEY));
+
+    if (ret == NULL) return NULL;
+
+    ret->references = 1;
+    return ret;
+}
+
+static void xor_freekey(void *keydata)
 {
-    return OPENSSL_zalloc(sizeof(XORKEY));
+    XORKEY* key = (XORKEY *)keydata;
+    int refcnt;
+
+    if (key == NULL)
+        return;
+
+    refcnt = atomic_fetch_sub_explicit(&key->references, 1,
+                                       memory_order_relaxed) - 1;
+    if (refcnt == 0)
+        atomic_thread_fence(memory_order_acquire);
+    if (refcnt > 0)
+        return;
+#ifndef NDEBUG
+    assert(refcnt == 0);
+#endif
+
+    if (key != NULL) {
+        OPENSSL_free(key->tls_name);
+	key->tls_name = NULL;
+    }
+    OPENSSL_free(key);
 }
 
-static void xor_freedata(void *keydata)
+static int xor_key_up_ref(XORKEY *key)
 {
-    OPENSSL_free(keydata);
+    int refcnt;
+
+    refcnt = atomic_fetch_add_explicit(&key->references, 1,
+                                       memory_order_relaxed) + 1;
```

Comment:
Use `CRYPTO_UP_REF` here

Reply:
so changed in https://github.com/openssl/openssl/pull/19312/commits/f716ba4cebb9c0586b746e99554b6f92aada9501